function runAPOBrareVariantAnalysis

% performs gene-based analysis with REGENIE for cirrhosis and HCC in the
% UKBB for meta-analysis with MVP and Milan cohorts.

clc

traits = ["Cirrhosis", "HCC"];

%% define cirrhosis and HCC (see PMID: 33347879) --------------------------
ex_codes = join(["B18" + (0:9), "B19" + (0:9)], ";");
phenoParser(query=["I850;I859;K703;K704;K721;K729;K741;K742;K746;K766;K767", ...
    "C220"],...
    tag=traits, ...
    all=true,...
    coding=repmat("19", 1, numel(traits)), ...
    ukbrap=true,...
    surv=false, ...
    remove=repmat(ex_codes, 1, numel(traits)));


%% run REGENIE step 1 to fit models ---------------------------------------
covars = ["Age", "Sex"];

step1_path = fullfile(pwd, "step1");
if ~isfolder(step1_path), mkdir(step1_path); end

for k = 1:numel(traits)

    out_dir = fullfile(step1_path, traits(k));
    if ~isfolder(out_dir), mkdir(out_dir); end

    call_regenie(step=1, ...
        bed="/dnax/GRM/ukb.qcn.indep.chr1-22.bed", ...
        trait=traits(k), ...
        covar=covars, ...
        agesexInt=true, ...
        threads=30, ...
        bsize=1000,...
        out=fullfile(out_dir, traits(k)), ...
        firth=true, approx=true, firth_se=true)
end

%% run rare variant analysis for each subset of APOB variants -------------
aaf_bins = 0.01; 
threads = 10;
genes = "APOB";

varset = load(fullfile(fileparts(pwd), "APOB.varset.mat"));
vep = varset.vep;
lc_lof_ids = vep.id(vep.LOFTEE == "LC"); % to remove based on LOFTEE predictions
varset = varset.apob;
fi = string(fieldnames(varset)); % APOB isoforms 

[covarFile, phenoFile, predfiles] = deal(strings(numel(traits), 1));
for k = 1:numel(traits)
    tmp = getfilenames(fullfile(step1_path, traits(k)), ".list", fullpath=true).list;
    predfiles(k) = tmp(tmp.endsWith("_pred.list"));

    tmp = getfilenames(fullfile(step1_path, traits(k)), ".txt", fullpath=true).txt;
    covarFile(k) = tmp(tmp.endsWith(".covarFile.txt"));
    phenoFile(k) = tmp(tmp.endsWith(".phenoFile.txt"));

end


for k = 1:numel(fi) % loop over isoforms (REGENIE step 2)
    rarevariants = setdiff(varset.(fi(k)), lc_lof_ids);
    
    pth_out = fullfile(pwd, "Results", "Overall", fi(k));
    if isfolder(pth_out), continue; end
    
    % gwasCaller by default activates firth correction before sending the
    % jobs to REGENIE
    % write_mask -> true for pQTL analysis

    gwasCaller(genes=genes, ...
        report=false, ...
        workers=threads, ...
        pheno_genes=repmat({traits},1,numel(genes)), ...
        pred=repmat({predfiles}, 1, numel(genes)), ...
        covarFile=repmat({covarFile}, 1, numel(genes)), ...
        phenoFile=repmat({phenoFile}, 1, numel(genes)), ...
        ignore_gene_scan=true, ...
        ignore_regional_plots=true, ...
        gene_scan_phewas=false, ...
        varset=rarevariants, ...
        aaf_bins=aaf_bins, ...
        dir=pth_out, ...
        weshome="/dnax/Bulk/Exome sequences/Population level exome OQFE variants, PLINK format - final release/", ...
        write_mask=true)

end

end % END